### PR TITLE
Exposing RoleAssignmentApproval APIs

### PR DESCRIPTION
**Pull request by shubhamguptacal**
[Overview of Role Assignment Approval Scenario](https://github.com/Azure/azure-rest-api-specs/files/5767066/RoleAssignmentApproval.pdf)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [ ] new API version 
      - [x] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [x] SDK of .NET (need service team to ensure code readiness)
      - [x] SDK of Python
      - [x] SDK of Java
      - [x] SDK of Js
      - [x] SDK of Go
      - [x] PowerShell
      - [x] CLI
      - [x] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [ ] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by yungezz** [comment]
Add `WaitForARMFeedback` since new APIs added

**Comment by chiragg4u** [comment]
Looks like there are many new API's introduced as part of this change. Please share the exact scenario and how these API's will be used and send it to armapireview alias for review to continue.

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json
Comment (yungezz) [inline specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:1063]: is approval Name supported?

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json
Comment (yungezz) [inline specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:2158]: in path, which param is approvalId? id? Is it supporting `name`? Since it's quite common to input name when querying resource

Comment (shubhamguptacal) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:2158]: In path, approval id needs to be passed in. The "name" is the same as approval Id

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json
Comment (shubhamguptacal) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:1063]: There is a "name" property on approval but it's the same as approvalId

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json
Comment (majastrz) [inline specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:1030]: >roleAssignmentApprovals [](start = 40, length = 23)

You are modifying a stable API version, which is not allowed as per guidance at [https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#versioning](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#versioning). You should add these changes in a new API version.

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json
Comment (majastrz) [inline specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:1033]: >ListRoleAssignmentApprovals [](start = 11, length = 27)

You should add paging for all these list calls. See [https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi009-always-add-x-ms-pageable-to-list-calls](https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi009-always-add-x-ms-pageable-to-list-calls)

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json
Comment (majastrz) [inline specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:1066]: I don't see any APIs to create approvals. I assume they created behind the scenes, but what actually triggers their creation? And how is it configured?

Comment (majastrz) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:1066]: And how does the chain of approvers get configured?

---
In reply to: [551591962](https://github.com/Azure/azure-rest-api-specs/pull/12108#discussion_r551591962) [](ancestors = 551591962)

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json
Comment (majastrz) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:2158]: To be more clear the `name` property should be the last segment of the value of the `id` property.

---
In reply to: [551552620](https://github.com/Azure/azure-rest-api-specs/pull/12108#discussion_r551552620) [](ancestors = 551552620)

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json
Comment (majastrz) [inline specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:2255]: >assignedToMe [](start = 9, length = 12)

Is this not read only?

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json
Comment (shubhamguptacal) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:1030]: Actually this API version was introduced very recently - last month(in this PR : https://github.com/Azure/azure-rest-api-specs/pull/10445). Is it okay to add these APIs in this version(since the version is very new) or do we always need to add new API versions for any chance?

Comment (shubhamguptacal) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:1030]: Also, this is not a stable version but a preview version

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json
Comment (shubhamguptacal) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:1033]: Thank you for pointing this out. I've fixed this.

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json
Comment (shubhamguptacal) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:1066]: Yes, approvals are created behind the scenes when a roleAssignmentRequest is created. I've added a PDF to the PR description which documents the flow and how approvals are created. I'll clarify more details in the meeting.

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json
Comment (shubhamguptacal) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json:2255]: Thank you for bringing this up. I've fixed this issue.

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json
Comment (shubhamguptacal) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2018-05-01-preview/authorization-AccessReviewCalls.json]: I've moved new APIs to a different version : 2021-01-01-preview. Thanks for bringing up this point!

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json
Comment (filizt) [inline specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json:66]: > "/providers/Microsoft.Authorization/roleAssignmentApprovals": { [](start = 3, length = 64)

We talked about "filter" being supported here. Please include a model for the filter options.

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json
Comment (filizt) [inline specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json:622]: Please mark "nextLink" as readOnly property.

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json
Comment (filizt) [inline specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json:778]: Nit - Origin has set of possible values. Consider making it an enum. [https://github.com/Azure/azure-resource-manager-rpc/blob/eec9f32c0357d0421be13c82ced767943013901a/v1.0/proxy-api-reference.md#origin-details](https://github.com/Azure/azure-resource-manager-rpc/blob/eec9f32c0357d0421be13c82ced767943013901a/v1.0/proxy-api-reference.md#origin-details)

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json
Comment (shubhamguptacal) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json:622]: Thanks for bringing this up! Fixed.

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json
Comment (filizt) [inline specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json:93]: >      "$ref": "#/definitions/ErrorDefinition" [](start = 8, length = 45)

Consider using common error model defined in types.json: [https://github.com/Azure/azure-rest-api-specs/blob/44ec11c74634f1afa5bbe0d3f2678b7f819ebf8e/specification/common-types/resource-management/v2/types.json](https://github.com/Azure/azure-rest-api-specs/blob/44ec11c74634f1afa5bbe0d3f2678b7f819ebf8e/specification/common-types/resource-management/v2/types.json). One example is : [https://github.com/Azure/azure-rest-api-specs/blob/18942e7ffaf9bc39da616397b4a6b8a191a00ea1/specification/maps/resource-manager/Microsoft.Maps/preview/2020-02-01-preview/maps-management.json#L88](https://github.com/Azure/azure-rest-api-specs/blob/18942e7ffaf9bc39da616397b4a6b8a191a00ea1/specification/maps/resource-manager/Microsoft.Maps/preview/2020-02-01-preview/maps-management.json#L88)

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json
Comment (shubhamguptacal) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json:778]: Thanks for bringing this up! I've documented all possible values

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json
Comment (shubhamguptacal) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json:66]: Thanks for bringing this up! I've added filter option parameter.

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json
Comment (shubhamguptacal) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json:96]: I copied this format from other APIs in Microsoft.Authorization RP. Do you think it's better to align with other APIs in the same RP?

Comment (majastrz) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json:96]: Filiz is right that the shared definition is the preferred approach, but for now let's align with the current APIs.

---
In reply to: [551663145](https://github.com/Azure/azure-rest-api-specs/pull/12108#discussion_r551663145) [](ancestors = 551663145)

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json
Comment (majastrz) [inline specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json:483]: >patch [](start = 7, length = 5)

PUT and PATCH verbs are both used for modifying an existing resource in ARM. (PUT is also used for creation, but that's not relevant in your case.) It's strange to support PATCH but not PUT.

You should add a PUT request for this as well. The difference is that PATCH allows a partial request body (usually just the property being modified) and PUT requires a full request body).

**Comment by majastrz** [comment]
Filiz and I have added some comments. Please take a look.

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json
Comment (shubhamguptacal) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json:537]: Thanks for bringing this up! Appreciate it! 

I've added support for PUT as well.

**Comment by shubhamguptacal** [comment]
Thanks @majastrz and @filizt for your feedback. Appreciate it! 

I've addressed your comments and updated the PR. Please take a look again.

File: specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json
Comment (majastrz) [reply specification/authorization/resource-manager/Microsoft.Authorization/preview/2021-01-01-preview/authorization-RoleAssignmentApprovalCalls.json:902]: Should model it using x-ms-enum. Take a look at [https://github.com/Azure/azure-rest-api-specs/blob/1907b81ca83cb8b8ad9438284b3f453d73aa3838/specification/common-types/resource-management/v2/types.json#L378](https://github.com/Azure/azure-rest-api-specs/blob/1907b81ca83cb8b8ad9438284b3f453d73aa3838/specification/common-types/resource-management/v2/types.json#L378).

---
In reply to: [551656722](https://github.com/Azure/azure-rest-api-specs/pull/12108#discussion_r551656722) [](ancestors = 551656722)

**Review by majastrz — APPROVED** [review]
Signed off from ARM side. There's a remaining minor comment. Please take a look.

**Comment by shubhamguptacal** [comment]
@yungezz can you please take a look again at this PR? Please don't merge it yet since we haven't deployed the APIs yet. I've added a donotmerge label(which I'll remove once we've deployed these APIs)

**Comment by yungezz** [comment]
> @yungezz can you please take a look again at this PR? Please don't merge it yet since we haven't deployed the APIs yet. I've added a donotmerge label(which I'll remove once we've deployed these APIs)

hi @shubhamguptacal I saw you removed `Do Not merge` label, does that mean the API change has been deployed to Prod? Pls ping me when the PR is ready to merge. thanks

File: specification/authorization/resource-manager/readme.md
Comment (yungezz) [inline specification/authorization/resource-manager/readme.md:262]: should the tag be package-2021-01-01-preview-only if only one swagger? could user complete an end to end scenario with only this one swagger file?

Comment (shubhamguptacal) [reply specification/authorization/resource-manager/readme.md:262]: Yes, for this specific API, the users are approvers of roleAssignmentApprovals. They can complete all actions necessary using the roleAssignmentApproval APIs. They don't have to deal with roleAssignment or any other entity

Comment (shubhamguptacal) [reply specification/authorization/resource-manager/readme.md]: Thanks for bringing this up! I've renamed to 2021-01-01-preview-only

**Comment by yungezz** [comment]
CI passed in ADO https://dev.azure.com/azure-sdk//internal/_build/results?buildId=738937&view=logs&j=2be18a23-047c-5c9f-61d7-60a33437d1f1, but the status isn't synced back to github.

**Review by yungezz — APPROVED** [review]

**Pull request merged**
